ui-lib/src/main/java/co/electriccoin/zcash/ui/common/FlowExt.kt

```diff
@@ -0,0 +1,52 @@
+package co.electriccoin.zcash.ui.common
+
+import co.electriccoin.zcash.spackle.Twig
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.withContext
+import kotlin.time.Duration
+import kotlin.time.ExperimentalTime
+import kotlin.time.TimeSource
+
+@OptIn(ExperimentalTime::class)
+fun <T> Flow<T>.throttle(
+    duration: Duration,
+    timeSource: TimeSource = TimeSource.Monotonic
+): Flow<T> = flow {
+    coroutineScope {
+        val context = coroutineContext
+        var timeMark = timeSource.markNow()
+        var delayEmit: Deferred<Unit>? = null
+        var firstValue = true
+        collect { value ->
+            if (firstValue) {
+                firstValue = false
+                emit(value)
+                timeMark = timeSource.markNow()
+                return@collect
+            }
+
+            if (timeMark.elapsedNow() >= duration) {
+                delayEmit?.cancel()
+                emit(value)
+                Twig.warn { " Normal Emmit $value" }
+                timeMark = timeSource.markNow()
+            } else {
+                delayEmit?.cancel()
+                delayEmit = async(Dispatchers.Default) {
+                    delay(duration /2)
```

Comment:
@ccjernigan after some testing, i found out that reducing delay to half the duration dramatically reduces the race condition. Without it the test would crash with throttle duration set to 0.01.seconds but with delay = duration / 2 i was able to go as low as 0.0001. It is probably not ideal but in theory "delayEmit" should be invoked only for the last item. What do you think?

Reply:
It sounds like you now have a test case that reproduces a race condition, which is good news!

Now that we can reproduce it, our goal isn't to reduce the race condition but to eliminate it entirely.  Code like this should be as deterministic as possible (knowing that we'll still have some timing variations just due to CPU clocks, etc).  So I don't recommend cutting the interval in half.

I'm not 100% certain but a Mutex seems like part of a possible solution.  E.g. The async block acquires the mutex to ensure exclusion, as well as the other branch of the if statement.